### PR TITLE
ENH: Arithmetic with Timestamp-based intervals

### LINKED ISSUE #35908: ENH: Arithmetic with Timestamp-based intervals
#### Is your feature request related to a problem?

It is currently possible to perform arithmetic on an `Interval` iff it is an `Interval` of numbers.  For `Interval` `iv`, `iv - c == Interval(iv.left-c, iv.right-c)`.  This does not work for intervals of `Timestamp`s:

```
In [90]: iv = pd.Interval(3, 4)

In [91]: pd.Interval(iv.left-2, iv.right-2)
Out[91]: Interval(1, 2, closed='right')

In [92]: iv - 2
Out[92]: Interval(1, 2, closed='right')

In [93]: iv = pd.Interval(pd.Timestamp("1900-01-01"), pd.Timestamp("1900-01-02"))

In [94]: pd.Interval(iv.left - pd.Timestamp("1900-01-01"), iv.right - pd.Timestamp("1900-01-0
    ...: 1"))
Out[94]: Interval(Timedelta('0 days 00:00:00'), Timedelta('1 days 00:00:00'), closed='right')

In [95]: iv - pd.Timestamp("1900-01-01")
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-95-b61ec78c0d66> in <module>
----> 1 iv - pd.Timestamp("1900-01-01")

TypeError: unsupported operand type(s) for -: 'pandas._libs.interval.Interval' and 'Timestamp'
```

#### Describe the solution you'd like

I would like that the behaviour that `iv - b == Interval(iv.left-b, iv.right-b)` holds for all types that `Interval` can hold, including `Timestamp`s and `Timedelta`s.

#### API breaking implications

No backward compatibility issues expected.

#### Describe alternatives you've considered

Users can currently resort to `Interval(iv.left-b, iv.right-b)`.  This is more typing and required repeating the expression `b`.

### PR DESCRIPTION
**Pull request by souris-dev**
- [X] closes #35908 
- [X] tests added / passed
- [X] passes `black pandas`
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v1.2.0.rst [hunk 1]
```diff
@@ -49,6 +49,42 @@ For example:
...
 
+Arithmetic with Timestamp and Timedelta-based Intervals
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Arithmetic can now be performed on :class:`Interval` s having their left and right
+ends as :class:`Timestamp`s or :class:`Timedelta` s, like what would be possible
+if the ends were numeric (:issue:`35908`).
+Before this change, it could be performed in an indirect way like this:
+
+.. ipython:: python
```
Comment (arw2019) [inline doc/source/whatsnew/v1.2.0.rst:60]: I feel like we don't need this here, the new method is better

Comment (souris-dev) [reply doc/source/whatsnew/v1.2.0.rst:60]: Sure, makes sense.

File: doc/source/whatsnew/v1.2.0.rst [hunk 2]
```diff
@@ -49,6 +49,42 @@ For example:
...
+Example:
+
+.. ipython:: python
+
+   interval = pd.Interval(pd.Timestamp("1900-01-01"), pd.Timestamp("1900-01-02"))
+   interval + pd.Timedelta("1 days 00:00:00")
+   interval - pd.Timedelta("1 days 00:00:00")
+
+However, it should be noted that :class:`Timestamp` s cannot be added, and a :class:`Timestamp`
+cannot be subtracted from a :class:`Timedelta`. Performing these would result in a
```
Comment (arw2019) [inline doc/source/whatsnew/v1.2.0.rst:85]: No need to say that it throws an error, just say that it's illegal

File: doc/source/whatsnew/v1.2.0.rst [hunk 3]
```diff
@@ -49,6 +49,42 @@ For example:
...
+ends as :class:`Timestamp`s or :class:`Timedelta` s, like what would be possible
+if the ends were numeric (:issue:`35908`).
+Before this change, it could be performed in an indirect way like this:
+
+.. ipython:: python
+
+   interval = pd.Interval(pd.Timestamp("1900-01-01"), pd.Timestamp("1900-01-02"))
+   pd.Interval(interval.left - pd.Timestamp("1900-01-01"), interval.right - pd.Timstamp("1900-01-01"))
+
+Now, this can be performed by directly using arithmetic operators (`-` or `+`):
```
Comment (arw2019) [inline doc/source/whatsnew/v1.2.0.rst:65]: Have just one example. We'll say at the top that this works for both Timestamps and Timedeltas

File: doc/source/whatsnew/v1.2.0.rst [hunk 4]
```diff
@@ -49,6 +49,42 @@ For example:
...
+
+.. ipython:: python
+
+   interval = pd.Interval(pd.Timestamp("1900-01-01"), pd.Timestamp("1900-01-02"))
+   interval - pd.Timestamp("1900-01-01")
+
+This is valid for addition using `+`, and also when the ends are :class:`Timedelta` s.
+Intervals having ends as Timestamps can also get "added to" or get "subtracted from",
```
Comment (arw2019) [inline doc/source/whatsnew/v1.2.0.rst:68]: Do we need this? Timestamps can't be added/subtracted even if they're not the bounds of an interval

Comment (souris-dev) [reply doc/source/whatsnew/v1.2.0.rst:68]: Timestamps can be subtracted but not added (addition makes no sense). Subtracting Timestamps give Timedeltas.

File: pandas/tests/scalar/interval/test_interval.py [hunk 5]
```diff
@@ -184,6 +184,184 @@ def test_math_sub(self, closed):
...
+        )
+
+        result = interval - Timedelta("3 days 01:00:00")
+        assert result == expected
+
+        result = interval
+        result -= Timedelta("3 days 01:00:00")
+        assert result == expected
+
+    def test_math_add_interval_timestamp_timestamp(self, closed):
```
Comment (arw2019) [inline pandas/tests/scalar/interval/test_interval.py:341]: `test_interval_addition_timestamp`
(left and right bounds have to be the same type)

File: pandas/tests/scalar/interval/test_interval.py [hunk 6]
```diff
@@ -184,6 +184,184 @@ def test_math_sub(self, closed):
...
+        )
+
+        msg = r"unsupported operand type\(s\) for \+"
+        with pytest.raises(TypeError, match=msg):
+            interval = interval + Timestamp("2002-01-08")
+
+        with pytest.raises(TypeError, match=msg):
+            interval += Timestamp("2002-01-08")
+
+    def test_math_sub_interval_timedelta_timestamp(self, closed):
```
Comment (arw2019) [inline pandas/tests/scalar/interval/test_interval.py:353]: `test_math_interval_mixed_timedelta_timestamp`

File: pandas/tests/scalar/interval/test_interval.py [hunk 7]
```diff
@@ -184,6 +184,184 @@ def test_math_sub(self, closed):
...
+        assert result == expected
+
+        expected = Interval(
+            interval.left - Timedelta("1 days 00:00:00"),
+            interval.right - Timedelta("1 days 00:00:00"),
+            closed=closed,
+        )
+        assert result == expected
+
+    def test_math_add_interval_timestamp_timedelta(self, closed):
```
Comment (arw2019) [inline pandas/tests/scalar/interval/test_interval.py:255]: `test_addition_interval_mixed_timestamp_timedelta`

File: pandas/tests/scalar/interval/test_interval.py [hunk 8]
```diff
@@ -184,6 +184,184 @@ def test_math_sub(self, closed):
...
+        )
+
+        result = interval + Timedelta("1 days 00:00:00")
+        assert result == expected
+
+        result = interval
+        result += Timedelta("1 days 00:00:00")
+        assert result == expected
+
+    def test_math_add_interval_timedelta_timedelta(self, closed):
```
Comment (arw2019) [inline pandas/tests/scalar/interval/test_interval.py:283]: `test_addition_interval_timedelta`

File: pandas/tests/scalar/interval/test_interval.py [hunk 9]
```diff
@@ -184,6 +184,184 @@ def test_math_sub(self, closed):
...
+        assert result == expected
+
+        expected = Interval(
+            interval.left - Timestamp("1900-01-01"),
+            interval.right - Timestamp("1900-01-01"),
+            closed=closed,
+        )
+        assert result == expected
+
+    def test_math_sub_interval_timestamp_timedelta(self, closed):
```
Comment (arw2019) [inline pandas/tests/scalar/interval/test_interval.py:221]: `test_subtraction_interval_mixed_timestamp_timedelta`

File: pandas/tests/scalar/interval/test_interval.py [hunk 10]
```diff
@@ -184,6 +184,184 @@ def test_math_sub(self, closed):
         with pytest.raises(TypeError, match=msg):
             interval - "foo"
 
+    def test_math_sub_interval_timestamp_timestamp(self, closed):
```
Comment (arw2019) [inline pandas/tests/scalar/interval/test_interval.py:187]: I'd name this `test_subtraction_interval_mixed_timestamp_timedelta`

**Review by arw2019 — CHANGES_REQUESTED** [review]
thanks @souris-dev for the PR!

Looks good, a few comments in the code

Re: tests generally I think you want to use `pytest.mark.parametrize` to make the tests easier to read

Re: docs you could add an example for the `pandas.Interval` API reference

Also, how does this work with Timestamps with different timezone info? Probably no need to mention in the docs but could add tests

**Comment by souris-dev** [comment]
@arw2019 - Thanks for the review!
I'll try incorporating these.
Also, the point you've raised about different timezones is interesting, I'll have a look at.

File: doc/source/whatsnew/v1.2.0.rst [hunk 11]
```diff
@@ -49,6 +49,26 @@ For example:
    buffer = io.BytesIO()
    data.to_csv(buffer, mode="w+b", encoding="utf-8", compression="gzip")
 
+Arithmetic with Timestamp and Timedelta-based Intervals
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Arithmetic can now be performed on :class:`Interval` s having their left and right
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v1.2.0.rst:55]: is there a reason for all the spaces between backtick and "s"?

File: pandas/_libs/interval.pyx [hunk 12]
```diff
@@ -413,6 +415,8 @@ cdef class Interval(IntervalMixin):
             isinstance(y, numbers.Number)
             or PyDelta_Check(y)
             or is_timedelta64_object(y)
+            or isinstance(y, _Timestamp)
+            or isinstance(y, _Timedelta)
```
Comment (jbrockmendel) [inline pandas/_libs/interval.pyx:419]: the PyDelta_Check call makes the _Timedelta check extraneous.  The remaining isinstance checks can be combined.

File: doc/source/whatsnew/v1.2.0.rst [hunk 13]
```diff
@@ -49,6 +49,26 @@ For example:
...
+ends as :class:`Timestamp` s or :class:`Timedelta` s, like what would be possible
+if the ends were numeric (:issue:`35908`).
+
+Arithmetic can be performed by directly using arithmetic operators (`-` or `+`),
+so something like this will work:
```
Comment (jreback) [inline doc/source/whatsnew/v1.2.0.rst:60]: can you add something like this as an example in Interval

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by arw2019 — COMMENTED** [review]
@souris-dev is this active?

**Comment by jreback** [comment]
closing as stale. if you want to continue, pls ping and can re-open.

**Pull request closed without merging**